### PR TITLE
adopt `torch.testing.assert_close` in test suite

### PR DESCRIPTION
**Pull request by pmeier**
We recently added [`torch.testing.assert_close`](), which allows us to compare all kind of numerics with expressive error messages if they don't match. This PR adopts it by going for the low hanging fruits: value comparisons of `torch.Tensor`'s or `np.ndarray`'s. We could also go for scalar or container of scalars comparisons, but IMO the `pytest` introspection is sufficient for this simple cases.

In contrast to the `torch.allclose` and `torch.eq` that we are currently using throughout the test suite, by default `torch.testing.assert_close` also checks for matching `dtype`'s, `device`'s and `stride()`'s. These checks can be relaxed by setting `check_(dtype|device|stride)=False`, but this should only be done, if we are sure that mismatching values there do not affect the test in any way. I've left these on by default which leads to a myriad of test failures. Although it means a lot of busy work, we should review these failures to see if we need to fix the test or relaxing the check is sufficient.

### CODE REVIEW

File: test/_assert_utils.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+"""This is a temporary module and should be removed as soon as torch.testing.assert_equal is supported."""
```
Comment (NicolasHug) [inline test/_assert_utils.py:1]: Maybe add a "TODO: remove this when..." comment so that we have a match when doing `git grep TODO`

**Review by NicolasHug — COMMENTED** [review]
Thanks a lot for working on this @pmeier 

> IMO the pytest introspection is sufficient for this simple cases. 

agreed

> I've left these on by default which leads to a myriad of test failures. Although it means a lot of busy work, we should review these failures to see if we need to fix the test or relaxing the check is sufficient

Do you need any help ATM or are you planning on fixing some of them on your own first?

Also, considering the amount of  current failures (~50), would it make sense to create PRs on a per-file basis? I fear that having everything in a single PR may increase the reviewing effort

**Comment by NicolasHug** [comment]
(CC @mthrok who might be interested)

File: test/test_image.py [hunk 2]
```diff
@@ -64,8 +65,7 @@ def test_decode_jpeg(self):
 
                 # Permit a small variation on pixel values to account for implementation
                 # differences between Pillow and LibJPEG.
-                abs_mean_diff = (img_ljpeg.type(torch.float32) - img_pil).abs().mean().item()
-                self.assertTrue(abs_mean_diff < 2)
+                torch.testing.assert_close(img_ljpeg, img_pil, rtol=0.0, atol=2.0)
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply test/test_image.py:68]: I'd be interested in having such fancier features in the future, and even semantics like "assert these 2 tensors are equal up to `tol`, except maybe on up to x% of their entries", to e.g. accept a few outliers

But regardless, in this particular case, I think the changes should be reverted: as you  already noted what is now tested is very different from the initial intended check, and it's not something that can be properly supported by `assert_close` at the moment.

**Comment by pmeier** [comment]
> Do you need any help ATM or are you planning on fixing some of them on your own first?

The fix is actually done fairly quickly since we only need to convert the dtype, device, or strides. IMO identifying what to do in every case is the harder part.

> Also, considering the amount of current failures (~50), would it make sense to create PRs on a per-file basis? I fear that having everything in a single PR may increase the reviewing effort

Maybe it is enough to "assign" reviewers to each file? IMO within each file with a multitude of failures we need to make only a few decision and apply that to every test in there. Otherwise I can split the PR.

**Comment by NicolasHug** [comment]
I think it'd be easier to have contained PRs so we can keep track of each decision independently

But to save time I'll see what I can do to fix `test_functional_tensor` (here)

**Comment by NicolasHug** [comment]
Switching to test_transforms now

**Comment by NicolasHug** [comment]
I tried to resolve most of the issues, hopefully there are less of them now. I'll stop here for tonight. I haven't looked at it in details but it looks like the GPU runs aren't using the latest pytest nightlies.

After working on this a bit I **really** think we should split this PR into multiple shorter ones now :)
Some of these changes can be a bit subtle and they deserve to be carefully reviewed. It will be easier to review in smaller PRs, and by separating them we'll be able to merge those that are obvious quicker.

**Comment by pmeier** [comment]
> After working on this a bit I **really** think we should split this PR into multiple shorter ones now :)

Since there are only 2 left, I'll do them here, Thanks a lot @NicolasHug!

**Comment by NicolasHug** [comment]
I created a bunch of sub-PRs (links above), hopefully I didn't miss any.
We can this one open and merge with master from time to time, until it becomes (hopefully)  empty.

File: test/_assert_utils.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+import functools
+
+import torch.testing
+
+__all__ = ["assert_equal"]
+
+
+assert_equal = functools.partial(torch.testing.assert_close, rtol=0, atol=0)
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/_assert_utils.py]: That's good feedback

**Comment by pmeier** [comment]
@NicolasHug this PR can be closed, correct?

**Pull request closed without merging**